pil-analyzer/src/type_inference.rs

```diff
@@ -526,21 +529,17 @@ impl<'a> TypeChecker<'a> {
                     .map(|item| self.infer_type_of_expression(item))
                     .collect::<Result<_, _>>()?,
             }),
-            Expression::LambdaExpression(
-                _,
-                LambdaExpression {
-                    kind: _,
-                    params,
-                    body,
-                },
-            ) => {
+            Expression::LambdaExpression(_, LambdaExpression { kind, params, body }) => {
                 let old_len = self.local_var_types.len();
                 let result = params
                     .iter()
                     .map(|p| self.infer_type_of_pattern(p))
                     .collect::<Result<Vec<_>, _>>()
                     .and_then(|param_types| {
-                        Ok((param_types, self.infer_type_of_expression(body)?))
+                        self.lambda_kind = Some(*kind);
+                        let body_type = self.infer_type_of_expression(body)?;
+                        self.lambda_kind = None;
```

Comment:
lambda expressions can be nested, so this should restore it to the previous value (just like with `local_var_types`)